File: nvpl_blas/c/zaxpy.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_zaxpy
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 5;
    nvpl_dcomplex_t * X;
    nvpl_dcomplex_t * Y;
    nvpl_dcomplex_t alpha = {1.0f, 2.0f};
    nvpl_int_t incX = 1;
    nvpl_int_t incY = 1;

    printf("\nExample: cblas_zaxpy for multiplying the vector x by the scalar and adds it to the vector y\n");
    printf("#### args: n=%" PRId64 ", incx=%" PRId64 ", incy=%" PRId64 ", alpha=(%g, %g)\n", (int64_t)N,
            (int64_t)incX, (int64_t)incY, alpha.real, alpha.imag);

    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    nvpl_int_t len_y = 1 + (N - 1) * labs(incY);
    // allocate memory
    X = (nvpl_dcomplex_t *)malloc(len_x * sizeof(nvpl_dcomplex_t));
    Y = (nvpl_dcomplex_t *)malloc(len_y * sizeof(nvpl_dcomplex_t));

    // fill data
    fill_zvector(X, N, incX);
    fill_zvector(Y, N, incY);

    // print input data
    print_zvector(X, N, incX, "X");
    print_zvector(Y, N, incY, "Y");

    // call cblas_zaxpy
    cblas_zaxpy(N, &alpha, X, incX, Y, incY);

    // print result
    printf("\nThe result of cblas_zaxpy:\n");
    print_zvector(Y, N, incY, "Y");

    // release memory
    free(X);
    free(Y);
    return 0;
}

File: nvpl_blas/f77/zaxpy.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     ZAXPY
!
!******************************************************************************
      program   ZAXPY_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      integer         n, incx, incy, i
      parameter       (n=5, incx=1, incy=1)
      double complex  alpha
      parameter       (alpha=1.0)
      integer         xsize, ysize
      parameter       (xsize=1+(n-1)*abs(incx),
     &                 ysize=1+(n-1)*abs(incy))
      double complex  x(xsize), y(ysize)
      data            (x(i), i=1,xsize)/1.0,-3.7,4.3,0.0,5.0/,
     &                (y(i), i=1,ysize)/ysize*2.0/

*       External Subroutines
      external        print_zvector
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, incx, incy, alpha
      call print_zvector(x,n,incx,'X')
      call print_zvector(y,n,incy,'Y')
*
*      CAll ZAXPY subroutine
      call ZAXPY(n,alpha,x,incx,y,incy)
*
      print*
      print 101
      call print_zvector(y,n,incy,'Y')

      stop
 99   format('Example: ZAXPY for multiplying the vector x by the',
     &       ' scalar and adds it to the vector y')
 100  format('#### args: n=',i1,', incx=',i1,', incy=',i1,
     &       ', alpha=(',f3.1,', ',f3.1,')')
 101  format('The result of ZAXPY: ')
      end
